Studio upload optimized workflows

wsmirnow opened the pull request:
The default opencast workflows have some issues processing video files created with opencast studio. As we also want an simplified workflow to process these files as fast as possible, I've created a new workflow to cover this case. The encoding profiles are optimized for performance and fixes the issues with uncommon video resolutions and variable frame rate.

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated testing
* [ ] have a clean commit history
* [ ] have proper commit messages (title and body) for all commits
* [ ] have appropriate tags applied

wsmirnow committed:
Fast studio workflow and encoding profiles created

The default opencast workflows have some issues processing video files created with opencast studio. As we also want an simplified workflow to process these files as fast as possible, I've created a new workflow to cover this case. The encoding profiles are optimized for performance and fixes the issues with uncommon video resolutions and variable frame rate.

wsmirnow committed:
Opencast Studio workflow set to studio-upload

lkiesow commented on etc/encoding/adaptive-streaming-movies.properties:
```diff
@@ -93,3 +93,45 @@ profile.adaptive-parallel.http.ffmpeg.command = -i #{in.video.path} \
+  #{if-height-geq-2160} \
+  #{if-height-geq-1440} \
+  #{if-height-geq-1080} \
+  -filter:v scale=w=1280:h='trunc(1280/dar/2)*2':force_original_aspect_ratio=disable,setsar=1/1,fps=fps=25 \
```
This is identical to the simpler

```suggestion
  -filter:v scale=w=1280:h='trunc(1280/dar/2)*2',setsar=1,fps=25 \
```

…though I'm wondering if it's better to use…

```
-filter:v scale=w=1280:720:force_original_aspect_ratio=decrease:force_divisible_by=2,fps=25 \
```
…which ensures that the video size does not exceed `1280:720`?
The same goes for the other filters.

lkiesow replied:
You included an `fps` filter in the pre-processing anyway, so there is no need to include it here again, is there?

wsmirnow replied:
The latest ffmpeg build from rpmfusion on fedora do not know the paramater `force_divisible_by` (anymore?!)

```
[Parsed_scale_0 @ 0x55c95e30efc0] Option 'force_divisible_by' not found
```
The way I made is complicatad but works on allmost every ffmpeg builds. Try to encode a video with an resolution of `1920x1058` without this option.

wsmirnow replied:
in this specific case we can remove the `fps` option.

lkiesow replied:
Checked and it's *not yet* included in the FFmpeg version in RPMFusion. So, it might make sense to wait a bit to introduce that and maybe make it a change for Opencast 9 later.

But I would still apply the suggestion above because that does not change the semantics of your command just the syntax and makes it easier to understand (less unnecessary calculations and no setting defaults)

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+
+  <id>studio-upload</id>
+  <title>Studio Upload</title>
+  <tags>
+    <tag>archive</tag>
+    <tag>upload</tag>
+  </tags>
```
If it's for studio, then it should not have these tags

```suggestion
```

wsmirnow replied:
> If it's for studio, then it should not have these tags

These tags are important for the case studio files should be ingested manually using download option of the video files in studio.

lkiesow replied:
I think it's confusing to have this additional workflow but we could remove the tags and keep the configuration panel. That would make it easy for users to make it visible if they want to.

lkiesow replied:
I was outvoted on this. Let's keep this as it is

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+            $('#publicationConfigurationPanel').toggle(show);
+            $('#straightToPublishing').val(show);
+            // Trigger update of the schedule's workflow config
+            $('#straightToPublishing').change();
+        });
+
+        <!-- called on loading an exisiting event -->
+        $('#straightToPublishing').bind('updateConfigUI', function() {
+            var publish = $(this).val().toLowerCase() === 'true';
+            $('#publicationConfigurationPanel').toggle(publish);
+            $('#holdOrPublishToggle_' + (publish ? 'publish':'hold')).prop('checked', 'checked');
+        });
+
+        <!-- Remove the line below if you wish to publish to AWS S3 -->
+        $('#publishToAws').prop('disabled', true);
+        <!-- Remove the line above if you wish to publish to AWS S3 -->
+
+        awsWasDisabled = $('#publishToAws').prop('disabled');
+        awsWasChecked = $('#publishToAws').prop('checked');
+        engageWasChecked = $('#publishToEngage').prop('checked');
+
+        $('input[name="publishToSearch"]').change(function(){
+            if ($(this).is(':checked')) {
+              $('#publishToAws').prop('checked', awsWasChecked);
+              $('#publishToEngage').prop('checked', engageWasChecked);
+              $('#publishToAws').prop('disabled', awsWasDisabled);
+              $('#publishToEngage').prop('disabled', false);
+            } else if (!$(this).is(':checked')) {
+              awsWasChecked = $('#publishToAws').prop('checked');
+              engageWasChecked = $('#publishToEngage').prop('checked');
+
+              $('#publishToAws').prop('disabled', true);
+              $('#publishToEngage').prop('disabled', true);
+              $('#publishToAws').prop('checked', false);
+              $('#publishToEngage').prop('checked', false);
+            }
+        });
+      </script>
+    ]]>
+  </configuration_panel>
```
This is unnecessary for studio as well

```suggestion
```

wsmirnow replied:
But neccessarry if we want to use the workflow to proccess studio videos manually upload by the Admin UI user.

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+        description="Tagging source material for archival">
+      <configurations>
+        <configuration key="source-flavors">dublincore/*,security/*,*/source</configuration>
+        <configuration key="target-tags">+archive</configuration>
```
It's a bit weird to add the tag to `*/source` here only to remove it later. You could even spare yourself this whole operation by just snapshotting the ACLs and dublincore catalogs according to their flavors later.

wsmirnow replied:
I know, it's weird. But in case some operation fail _before_ we take the first snapshot for some reason, you will be save. After you've sucessfully transcoded the source material, the flavor (and tag) will be switched.

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+        <configuration key="workflow-id">partial-transcode-studio-tracks</configuration>
+      </configurations>
+    </operation>
+
+    <operation
+        id="snapshot"
+        fail-on-error="true"
+        exception-handler-workflow="partial-error"
+        description="Archiving">
+      <configurations>
+        <configuration key="source-tags">archive</configuration>
+      </configurations>
+    </operation>
```
Why create a snapshot here? You create another one later anyway.

```suggestion
```

wsmirnow replied:
Cause of loosing the data if some operation later fail.

lkiesow commented on etc/encoding/opencast-movies.properties:
```diff
@@ -168,3 +168,19 @@ profile.import.silent.input = nothing
 profile.import.silent.ffmpeg.command = -strict -2 -filter_complex aevalsrc=0::d=#{time} -c:a aac -b:a 8k #{out.dir}/#{out.name}#{out.suffix}
+
+# Transcode videos generated with opencast studio to cover some issues
+# The video resolution should be reduced to maximum of 1080p.
```
It makes no sense to limit this to 1080p here but then include 2.5k and 4k encoding later.

lkiesow commented on etc/encoding/opencast-movies.properties:
```diff
@@ -168,3 +168,19 @@ profile.import.silent.input = nothing
+profile.scale.fhd.framerate.fix.suffix = .mp4
+profile.scale.fhd.framerate.fix.ffmpeg.command = -i #{in.video.path} \
+  -max_muxing_queue_size 1000 \
+  -filter:v scale=w='trunc(min(max(1920\\,in_w)\\,in_w)/2)*2':'trunc(out_w/dar/2)*2':force_original_aspect_ratio=disable,setsar=1/1,fps=fps=25 \
```
This should be identical to
```suggestion
  -filter:v scale=w='trunc(min(max(1920\\,in_w)\\,in_w)/2)*2':'trunc(out_w/dar/2)*2':setsar=1,fps=25 \
```
<strike>
but then again, wouldn't it be easier to use
```
 -filter:v scale=1920:1080:force_original_aspect_ratio=decrease:force_divisible_by=2,fps=25
```
</strike>

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+    </operation>
+
+    <operation
+        id="prepare-av"
```
Just wondering if you couldn't do the initial re-encoding in here instead of adding another `encode` operation?

wsmirnow replied:
This operations will be skipped in most cases. Opencast studio pack the audio stream in both video files on dualstream recording. But just in case…

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+        description="Creating player preview image">
+      <configurations>
+        <configuration key="source-flavor">*/prepared</configuration>
+        <configuration key="source-tags"></configuration>
```
```suggestion
```

wsmirnow replied:
This is an copy paste artifact from existing workflow I've copied from, see https://github.com/opencast/opencast/blob/5144e24f9604df2f9aa3609c0936958c91e96f09/etc/workflows/partial-publish.xml#L157
We can remove it.

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+
+    <!-- Create a cover image with the default template -->
+
+    <operation
+        id="tag"
+        description="Removing unneeded presenter-cover from download publication">
+      <configurations>
+        <configuration key="source-flavors">presenter/player+preview</configuration>
+        <configuration key="target-tags">-engage-download</configuration>
+      </configurations>
+    </operation>
```
Why do you create the image in the first place if you remove the tag here so it's never being used?

wsmirnow replied:
Same as before, copy paste issue. But the question is, does the player show the presenter preview image?

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+        description="Remove temporary processing artifacts">
+      <configurations>
+        <configuration key="preserve-flavors">security/*</configuration>
+        <configuration key="delete-external">false</configuration>
```
This makes a lot of sense in a lot of cases but is not a safe choice in every case
```suggestion
        <configuration key="delete-external">true</configuration>
```

wsmirnow replied:
The not save case is not really exists on studio uploads.

lkiesow commented on etc/workflows/studio-upload.xml:
```diff
@@ -0,0 +1,448 @@
+        <configuration key="publishToApi">true</configuration>
+        <configuration key="publishToOaiPmh">false</configuration>
+        <configuration key="publishToAws">false</configuration>
+        <configuration key="thumbnailType">0</configuration>
+        <configuration key="thumbnailPosition">1</configuration>
```
What are these for?

wsmirnow replied:
Is the question about the last two properties? They will be expected by the partial-preview (existing) workflow.

lkiesow reviewed (changes requested):
This is a large step in the right direction even for the default workflow.
I still found a few issues ;-P

wsmirnow commented:
I see this patch as a quick fix for studio videos. I'm mostly done with refactoring the default workflows. It will be part of an other PR in develop. I hope this quickfix workflow will not be needed there any more.

wsmirnow committed:
Apply suggestions from code review

Review fixes

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

lkiesow commented:
Upon discussing this a bit further we found that most of my complaints were already present in the default workflows and it makes more sense to keep them similar for now and then release a larger update for Opencast 9 (hopefully in the next days).

wsmirnow committed:
adaptive encoding profile optimization

CGreweling commented:
When I try this i get:
setsar not found.


```
Executing encoding command: [ffmpeg, -nostdin, -nostats, -i, /media/opencaststorage/raw/oc-test-worksp
ace/https_oc-test.virtuos.uni-osnabrueck.de/assets/assets/b935882e-c91f-4cb0-a6e3-fb87a1532092/556cecfb-c76d-44bc-b28b-9e1b75b4bbc0/6/oc_studio_2020_03_20_16_
41_presenter.mkv, -max_muxing_queue_size, 1000, -filter:v, scale=w=trunc(min(max(1920\,in_w)\,in_w)/2)*2:trunc(out_w/dar/2)*2:setsar=1,fps=25, -c:a, aac, -ab,
 128k, -c:v, libx264, -crf, 21, -movflags, +faststart, /media/opencaststorage/raw/oc-test-workspace/https_oc-test.virtuos.uni-osnabrueck.de/assets/assets/b935
882e-c91f-4cb0-a6e3-fb87a1532092/556cecfb-c76d-44bc-b28b-9e1b75b4bbc0/6/oc_studio_2020_03_20_16_41_presenter_a8178145-d752-4002-a2aa-2ba14e2b64cd.mp4]
2020-03-21T16:15:49,662 | INFO  | (EncoderEngine:461) - [Parsed_scale_0 @ 0x4785a40] Option 'setsar' not found
```

lkiesow committed:
Fix Typo In Encoding Profile

This patch fixes a typo in the filter definition of the encoding profile
used in the studio workflow.

lkiesow commented:
@CGreweling, ups, that was a typo we probably introduced when updating the profiles Friday late. Sorry. I took the liberty of fixing it. Can you try again?

@gregorydlogan, the workflow should be ready to go, with the small caveats we discussed on Friday: Thigs can be improved but this tries to stay close to `schedule-and-upload`.

CGreweling commented:
@lkiesow works now thanks. 
One Idea. Why not put the custom encoding profiles in one studio.properties.
That would make maintenance a little easier?
